## House Pricing Assignment 2

### Shivpreet Padhi

In [1]:
# Imports
import sklearn
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Loading Data
df = pd.read_csv('housing dataset.csv')
df = df.dropna()

In [3]:
df.head()

MSSubClass  LotFrontage  LotArea  OverallQual  WoodDeckSF  OpenPorchSF  \
0          60         65.0     8450            7           0           61   
1          20         80.0     9600            6         298            0   
2          60         68.0    11250            7           0           42   
3          70         60.0     9550            7           0           35   
4          60         84.0    14260            8         192           84   

   TotalBsmtSF   FloorSF  SalePrice  
0          856  104678.0     208500  
1         1262   91381.0     181500  
2          920  112210.0     223500  
3          756   70378.0     140000  
4         1145  125572.5     250000

In [4]:
# Select featues and target variables
X = df[['MSSubClass', 'LotFrontage', 'LotArea','OverallQual','WoodDeckSF','OpenPorchSF','FloorSF','TotalBsmtSF']]
Y = df['SalePrice']

In [5]:
# Split data into train and test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=0)

In [6]:
#Training
from sklearn.linear_model import LinearRegression
regressor=LinearRegression()
regressor.fit(x_train,y_train)

LinearRegression()

In [7]:
#intercept and coff
print(regressor.intercept_)
print(regressor.coef_)

5.820766091346741e-11
[-4.93855407e-13 -1.78207146e-12  8.05932580e-16  1.55994653e-11
  5.18095943e-15  3.83830109e-14  2.00000000e+00 -1.00000000e+00]


In [8]:
# prediction
y_predict=regressor.predict(x_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_predict})

In [9]:
y_train_predict = regressor.predict(x_train)

In [10]:
#Accuracy of train data
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_train, y_train_predict))  
print('Mean Squared Error:', metrics.mean_squared_error(y_train, y_train_predict))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_train, y_train_predict)))

Mean Absolute Error: 3.48791218129918e-11
Mean Squared Error: 2.7241814840079713e-21
Root Mean Squared Error: 5.219369199441606e-11


In [11]:
#Accuracy of test data
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_predict))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_predict))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_predict)))

Mean Absolute Error: 3.269652322058361e-11
Mean Squared Error: 2.1199986965641217e-21
Root Mean Squared Error: 4.6043443578474034e-11


# RandomForestRegressor

In [12]:
from sklearn.ensemble import RandomForestRegressor

In [13]:
rf=RandomForestRegressor(n_estimators=500, max_depth=10)
rf.fit(x_train,y_train)

RandomForestRegressor(max_depth=10, n_estimators=500)

In [14]:
# prediction
y_predict=rf.predict(x_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_predict})
df

Actual      Predicted
242    79000   78922.395556
400   245500  244886.134000
54    130000  129453.432434
316   260000  259677.372095
748   260400  259985.816000
...      ...            ...
65    317000  315626.298198
601   141000  141072.668750
826   109500  109463.706661
1099  157000  157202.882378
1052  165000  165026.672170

[241 rows x 2 columns]

In [15]:
y_train_predict = rf.predict(x_train)
#Accuracy of train data
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_train, y_train_predict))  
print('Mean Squared Error:', metrics.mean_squared_error(y_train, y_train_predict))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_train, y_train_predict)))

Mean Absolute Error: 325.0502403702287
Mean Squared Error: 4537660.7983521605
Root Mean Squared Error: 2130.1785836760637


In [16]:
#Accuracy of test data
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_predict))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_predict))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_predict)))

Mean Absolute Error: 428.92995558562666
Mean Squared Error: 645337.8269378048
Root Mean Squared Error: 803.3292145427084


## Gradent Boosting Regressor

In [17]:
from sklearn import ensemble

In [18]:
reg = ensemble.GradientBoostingRegressor()
reg.fit(x_train, y_train)

GradientBoostingRegressor()

In [19]:
# prediction
y_predict=reg.predict(x_test)
df = pd.DataFrame({'Actual': y_test, 'Predicted': y_predict})
df

Actual      Predicted
242    79000   79065.690656
400   245500  248186.911147
54    130000  129304.133257
316   260000  259507.090890
748   260400  259850.580255
...      ...            ...
65    317000  316484.520452
601   141000  141851.366601
826   109500  109430.272190
1099  157000  156971.224871
1052  165000  164838.544451

[241 rows x 2 columns]

In [20]:
y_train_predict = reg.predict(x_train)
#Accuracy of train data
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_train, y_train_predict))  
print('Mean Squared Error:', metrics.mean_squared_error(y_train, y_train_predict))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_train, y_train_predict)))

Mean Absolute Error: 422.6550797033422
Mean Squared Error: 347301.6675684626
Root Mean Squared Error: 589.3230587449151


In [21]:
#Accuracy of test data
from sklearn import metrics
print('Mean Absolute Error:', metrics.mean_absolute_error(y_test, y_predict))  
print('Mean Squared Error:', metrics.mean_squared_error(y_test, y_predict))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y_test, y_predict)))

Mean Absolute Error: 725.8446347455434
Mean Squared Error: 1056834.7302747425
Root Mean Squared Error: 1028.0246739620322


## Final Errors
(Mean Absolute)

### Linear Regressor
Training: 3.48791218129918e-11

Test: 3.269652322058361e-11

### Random Forest
Training: 325.0502403702287

Test: 428.92995558562666

### Gradient Boosting
Training: 422.6550797033422

Test: 725.8446347455434

## Conclusion
The best result is obtained from Linear Regressor algorithm. Random forest and gradent boosting donot give appropriate results. Hence, for this dataset linear regressor is the best match.